File: library/src/main/java/net/robinx/queue/DefaultDelivery.java
package net.robinx.queue;

import android.os.Handler;

import java.util.List;
import java.util.concurrent.Executor;

/**
 * Created by Robin on 2016/12/6 15:44.
 */

public class DefaultDelivery<T extends Job<T>> implements Delivery<T> {
    private final Executor mResponsePoster;

    public DefaultDelivery(final Handler handler) {
        mResponsePoster = new Executor() {
            @Override
            public void execute(Runnable command) {
                handler.post(command);
            }
        };
    }

    public DefaultDelivery(Executor executor) {
        mResponsePoster = executor;
    }

    @Override
    public void deliveryPrepare(final List<JobHandlerListener<T>> listeners, final T job) {
        mResponsePoster.execute(new Runnable() {
            @Override
            public void run() {
                if (job.isCancel()) {
                    CLog.e("job is cancel when deliveryPrepare");
                    deliveryCancel(listeners,job);
                    return;
                }

                CLog.d("deliveryPrepare,current thread:%s", Thread.currentThread().getName());
                for (int i = 0, size = listeners.size(); i < size; i++) {
                    listeners.get(i).onPrepare(job);

                }
            }
        });
    }

    @Override
    public void deliveryCancel(final List<JobHandlerListener<T>> listeners, final T job) {
        mResponsePoster.execute(new Runnable() {
            @Override
            public void run() {
                CLog.d("deliveryCancel,current thread:%s", Thread.currentThread().getName());
                for (int i = 0, size = listeners.size(); i < size; i++) {
                    listeners.get(i).onCancel(job);
                }
            }
        });
    }

    @Override
    public void deliveryFinish(final List<JobHandlerListener<T>> listeners, final T job) {
        if (job.isCancel()) {
            CLog.e("job is cancel when deliveryFinish");
            deliveryCancel(listeners,job);
            return;
        }

        mResponsePoster.execute(new Runnable() {
            @Override
            public void run() {
                CLog.d("deliveryFinish,current thread:%s", Thread.currentThread().getName());
                for (int i = 0, size = listeners.size(); i < size; i++) {
                    listeners.get(i).onFinish(job);
                }
            }
        });
    }
}

File: library/src/main/java/com/github/jobqueue/DefaultDelivery.kt
package com.github.jobqueue

import android.os.Handler
import java.util.concurrent.Executor

/**
 * Created by Robin on 2016/12/6 15:44.
 */
class DefaultDelivery<T : Job<T>> : Delivery<T> {
    private val mResponsePoster: Executor

    constructor(handler: Handler) {
        mResponsePoster = Executor { command -> handler.post(command) }
    }

    constructor(executor: Executor) {
        mResponsePoster = executor
    }

    override fun deliveryPrepare(listeners: List<JobHandlerListener<T>?>?, job: T) {
        mResponsePoster.execute(Runnable {
            if (job.isCancel) {
                CLog.e("job is cancel when deliveryPrepare")
                deliveryCancel(listeners, job)
                return@Runnable
            }
            CLog.d("deliveryPrepare,current thread:%s", Thread.currentThread().name)
            var i = 0
            val size = listeners!!.size
            while (i < size) {
                listeners[i]!!.onPrepare(job)

                i++
            }
        })
    }

    override fun deliveryCancel(listeners: List<JobHandlerListener<T>?>?, job: T) {
        mResponsePoster.execute {
            CLog.d("deliveryCancel,current thread:%s", Thread.currentThread().name)
            var i = 0
            val size = listeners!!.size
            while (i < size) {
                listeners[i]!!.onCancel(job)
                i++
            }
        }
    }

    override fun deliveryFinish(listeners: List<JobHandlerListener<T>?>?, job: T) {
        if (job.isCancel) {
            CLog.e("job is cancel when deliveryFinish")
            deliveryCancel(listeners, job)
            return
        }

        mResponsePoster.execute {
            CLog.d("deliveryFinish,current thread:%s", Thread.currentThread().name)
            var i = 0
            val size = listeners!!.size
            while (i < size) {
                listeners[i]!!.onFinish(job)
                i++
            }
        }
    }
}
